### PR TITLE
Fix query based dropdown not adding quote marks correctly

### LINKED ISSUE #4184: Query with a drop down list based on query (Allow multiple values with quotation mark) not adding quotation mark correctly
### Issue Summary

Query with a drop down list based on query (Allow multiple values with quotation mark) not adding quotation mark correctly.

### Steps to Reproduce

1. Create a query with a parameter with query based dropdown list.
2. Select allow multiple values, double quotation mark.
3. Execute query, when the parameter select only 1 choice and the parameter is text, it didn't add double quotation mark and the query returns an error.

### Technical details:

* Redash Version: 8.0 Beta 2
* Browser/OS: CentOS 7
* How did you install Redash: Docker

### PR DESCRIPTION
**Pull request by gabrieldutra**
## What type of PR is this? (check all applicable)

- [x] Bug Fix

## Description
This is not perfect yet, I still wanna explore a few more options without having the Dropdown Options in the Parameter structure context (that's the only difference from the regular Dropdown code - what handles if the value is valid is the Input).

## Related Tickets & Documents
Fixes #4184

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)
--

### CODE REVIEW

File: client/app/services/query.js [hunk 1]
```diff
@@ -387,7 +393,8 @@ export class Parameter {
       if (has(query, key)) {
         if (this.multiValuesOptions) {
           try {
-            this.setValue(JSON.parse(query[key]));
+            const valueFromJson = JSON.parse(query[key]);
+            this.setValue(isArray(valueFromJson) ? valueFromJson : query[key]);
```
Comment (gabrieldutra) [inline client/app/services/query.js]: Turns out numbers are also parsed and converted :sweat_smile:, so this makes sure we don't convert it unless it's an array.

File: client/app/services/query.js [hunk 2]
```diff
@@ -253,6 +253,12 @@ export class Parameter {
       }
     }
 
+    if (this.type === 'query' && !isEmptyValue(value)) {
+      if (this.multiValuesOptions && !isArray(value) && value !== null) {
+        value = [value];
```
Comment (gabrieldutra) [inline client/app/services/query.js]: QueryBasedDropdown with multi-values should always hold array values in its structure.

File: client/cypress/integration/query/parameter_spec.js [hunk 3]
```diff
@@ -180,6 +180,69 @@ describe('Parameter', () => {
...
+            ],
+          },
+        };
+
+        cy.visit(`/queries/${dropdownQuery.id}`);
+        cy.getByTestId('ExecuteButton').click();
+        cy.getByTestId('TableVisualization')
+          .should('contain', 'value1')
+          .and('contain', 'value2')
+          .and('contain', 'value3');
```
Comment (ranbena) [inline client/cypress/integration/query/parameter_spec.js]: Better move this to a test of its own? (I don't think setup is a right place for assertions)

File: client/app/services/query.js [hunk 4]
```diff
@@ -253,6 +253,12 @@ export class Parameter {
       }
     }
 
+    if (this.type === 'query' && !isEmptyValue(value)) {
+      if (this.multiValuesOptions && !isArray(value) && value !== null) {
```
Comment (ranbena) [inline client/app/services/query.js]: Don't think you need to check `value !== null` cause it's checked the line before.

File: client/cypress/integration/query/parameter_spec.js [hunk 5]
```diff
@@ -180,6 +180,69 @@ describe('Parameter', () => {
...
+      cy.getByTestId('QueryEditor').click(); // just to close the select menu
+
+      cy.getByTestId('ParameterApplyButton')
+        .click();
+
+      cy.getByTestId('TableVisualization')
+        .should('contain', '"1","2","3"');
+    });
+  });
+
```
Comment (ranbena) [inline client/cypress/integration/query/parameter_spec.js]: Will you add a #4184 specific test?

File: client/cypress/integration/query/parameter_spec.js (hunk 3 above)
Comment (gabrieldutra) [reply client/cypress/integration/query/parameter_spec.js]: Considering Cypress style I don't see any problems on assertions in the setup (they are quite fast, new tests are not). But in this case I wanted to make sure the query was executed and had results (needed for the Dropdown), so the assertion is indeed part of the setup.

File: client/cypress/integration/query/parameter_spec.js (hunk 5 above)
Comment (gabrieldutra) [reply client/cypress/integration/query/parameter_spec.js]: I thought about it, but a general functionality test seemed more efficient as we cover more cases. In any case, this test fails in master and passes in here.

File: client/app/services/query.js (hunk 4 above)
Comment (gabrieldutra) [reply client/app/services/query.js]: Yeah, I was a bit paranoid on this one :sweat_smile:

**Comment by arikfr** [comment]
:bug:

**Pull request merged**